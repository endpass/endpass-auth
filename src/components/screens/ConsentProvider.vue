<template>
  <v-frame>
    <scopes-form
      :error="error"
      :loading="loading"
      :scopes="scopes"
      @submit="handleScopesSubmit"
    />
  </v-frame>
</template>

<script>
import get from 'lodash/get';
import { mapActions, mapState } from 'vuex';
import { queryParamsToObject } from '@/util/url';
import VFrame from '@/components/common/VFrame';
import ScopesForm from '@/components/forms/Scopes';

export default {
  name: 'ConsentProvider',

  data: () => ({
    params: {},
    error: null,
  }),

  computed: {
    ...mapState({
      loading: state => state.core.loading,
      isAuthorized: state => state.accounts.isAuthorized,
    }),

    scopes() {
      const scopes = get(this.params, 'scopes', '');

      if (scopes) {
        return decodeURIComponent(scopes).split(' ');
      }

      return [];
    },
  },

  methods: {
    ...mapActions(['grantPermissionsWithHydra']),

    async handleScopesSubmit(scopes) {
      const { consentChallenge } = this.params;

      if (!consentChallenge) return;

      try {
        await this.grantPermissionsWithHydra({ consentChallenge, scopes });
      } catch (err) {
        this.error = err.message;
      }
    },
  },

  mounted() {
    const { search } = window.location;

    this.params = queryParamsToObject(search);

    if (!this.params.consentChallenge) {
      this.error =
        'You should provide consent_challenge param in url, add it and try again!';
      return;
    }

    if (!this.params.scopes) {
      this.error =
        'You should provide scopes param in url, add it and try again!';
      return;
    }

    if (!this.isAuthorized) {
      this.$router.replace(`/public/auth${search}&place=consent`);
    }
  },

  components: {
    VFrame,
    ScopesForm,
  },
};
</script>
