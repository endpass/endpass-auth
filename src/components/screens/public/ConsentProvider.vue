<template>
  <v-frame :closable="false">
    <v-error
      v-if="error.show"
      :hint="error.hint"
      :description="error.description"
    />
    <scopes-form
      v-else
      :loading="isLoading"
      :scopes-list="scopesList"
      @submit="handleScopesSubmit"
    />
  </v-frame>
</template>

<script>
/* eslint-disable camelcase */

import get from 'lodash/get';
import { mapActions, mapState } from 'vuex';
import VFrame from '@/components/common/VFrame';
import ScopesForm from '@/components/forms/Scopes';
import VError from '@/components/common/VError';
import Spinner from '@/components/common/Spinner';

export default {
  name: 'ConsentProvider',

  data: () => ({
    queryParamsMap: {},
    initialHref: window.location.href,
    isLoading: true,
    scopesList: [],
    error: {
      show: false,
      hint: '',
      description: '',
    },
  }),

  computed: {
    ...mapState({
      inited: state => state.core.inited,
      isLogin: state => state.accounts.isLogin,
    }),
  },

  methods: {
    ...mapActions(['grantPermissionsWithHydra', 'getConsentDetails']),

    async handleScopesSubmit(scopesList) {
      this.isLoading = true;
      try {
        const { redirect } = await this.grantPermissionsWithHydra({
          consentChallenge: this.queryParamsMap.consent_challenge,
          scopesList,
        });
        window.location.href = redirect;
      } catch (err) {
        this.setError(err.message);
      }
      this.isLoading = false;
    },

    setError(hint, description = '') {
      this.error = {
        show: true,
        hint,
        description,
      };
    },

    async loadScopes() {
      this.isLoading = true;

      try {
        // load scopes
        const { requested_scope } = await this.getConsentDetails(
          this.queryParamsMap.consent_challenge,
        );
        this.scopesList = requested_scope;
      } catch (e) {
        this.setError('Something broken, when loading scopes');
      }
      this.isLoading = false;
    },
  },

  async mounted() {
    const { query } = get(this.$router, 'history.current', {});

    this.queryParamsMap = query;

    if (query.error) {
      this.setError(query.error_hint, query.error_description);
      return;
    }

    if (!query.consent_challenge) {
      this.setError(
        'You should provide consent_challenge param in url, add it and try again!',
      );
      return;
    }

    if (!this.isLogin) {
      this.$router.replace(
        `/public/auth?redirect_url=${encodeURI(
          this.initialHref,
        )}&place=consent`,
      );
      return;
    }

    await this.loadScopes();
  },

  components: {
    Spinner,
    VError,
    VFrame,
    ScopesForm,
  },
};
</script>
