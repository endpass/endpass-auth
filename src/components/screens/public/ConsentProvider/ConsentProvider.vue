<template>
  <loading-screen v-if="isLoadingScreen" />
  <v-frame
    v-else
    :title="$t('components.consentProvider.title')"
    :is-closable="isClosable"
    :is-show-logo="true"
    @close="onClose"
  >
    <v-error
      v-if="isError"
      :label="errorLabel"
      :description="errorDescription"
    />
    <scopes-form
      v-if="scopesList.length > 0"
      :is-loading="isLoading"
      :scopes-list="scopesList"
      @submit="handleScopesSubmit"
    />
  </v-frame>
</template>

<script>
import LoadingScreen from '@/components/common/LoadingScreen';
import VFrame from '@/components/common/VFrame';
import ScopesForm from '@/components/forms/Scopes';
import VError from '@/components/common/VError';
import { authStore, coreStore } from '@/store';
import createConsentProviderController from './ConsentProviderController';

export default {
  name: 'ConsentProvider',

  coreStore,
  authStore,

  consentProviderController: createConsentProviderController(),

  data: () => ({
    consentChallenge: null,
    isLoading: true,
    isSkipped: false,
    scopesList: [],
    errorHint: '',
    errorDescription: '',
  }),

  computed: {
    isError() {
      return !!this.errorLabel;
    },

    errorLabel() {
      return this.errorHint || this.$options.consentProviderController.error;
    },

    isClosable() {
      return this.$options.coreStore.isDialog;
    },

    isLogin() {
      return this.$options.authStore.isLogin;
    },

    isLoadingScreen() {
      return this.isSkipped || (this.scopesList.length === 0 && !this.isError);
    },
  },

  methods: {
    async handleScopesSubmit(scopesList) {
      try {
        this.isLoading = true;
        await this.$options.consentProviderController.grantScopes({
          consentChallenge: this.consentChallenge,
          scopesList,
        });
      } catch (e) {
        this.errorHint = e.message;
      } finally {
        this.isLoading = false;
      }
    },

    onClose() {
      this.$options.coreStore.cancelAllChannels();
      this.$options.coreStore.dialogClose();
    },

    async loadScopes() {
      try {
        this.isLoading = true;

        const {
          scopesList,
          isSkip,
        } = await this.$options.consentProviderController.loadScopes({
          consentChallenge: this.consentChallenge,
        });

        this.isSkipped = isSkip;
        this.scopesList = scopesList;
      } catch (e) {
        this.errorHint = e.message;
      } finally {
        this.isLoading = false;
      }
    },
  },

  async mounted() {
    const { query } = this.$route;
    const { consent_challenge: consentChallenge, error } = query;

    this.consentChallenge = consentChallenge;

    if (error) {
      const {
        error_hint: errorHint,
        error_description: errorDescription,
      } = query;

      this.errorHint = errorHint;
      this.errorDescription = errorDescription;
      return;
    }

    if (!consentChallenge) {
      this.errorHint = this.$i18n.t('components.consentProvider.urlError');
      return;
    }

    if (!this.isLogin) {
      this.$router.replace({
        name: 'PublicAuthScreen',
        query: {
          redirectUrl: encodeURI(window.location.href),
          place: 'consent',
        },
      });
      return;
    }

    await this.loadScopes();
  },

  components: {
    LoadingScreen,
    VError,
    VFrame,
    ScopesForm,
  },
};
</script>
