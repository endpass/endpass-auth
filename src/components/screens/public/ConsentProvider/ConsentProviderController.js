import { VuexModule, Action, Module } from 'vuex-class-modules';
import createController from '@/controllers/createController';
import { accountsStore as accountsStoreModule } from '@/store';
import i18n from '@/locales/i18n';

const OPEN_URL_TIMEOUT_MS = 5000;

@Module({ generateMutationSetters: true })
class ConsentProviderController extends VuexModule {
  error = '';

  constructor(props, { accountsStore = accountsStoreModule }) {
    super(props);
    this.accountsStore = accountsStore;
  }

  openRedirectUrl(url) {
    window.location.href = url;
    setTimeout(() => {
      this.error = i18n.t('components.consentProvider.redirectUrl');
    }, OPEN_URL_TIMEOUT_MS);
  }

  /**
   * @param {object} params
   * @param {String[]} params.scopesList
   * @return {Promise<void>}
   */
  @Action
  async submitScopes({ consentChallenge, scopesList }) {
    try {
      const { redirect } = await this.accountsStore.grantPermissionsWithOauth({
        consentChallenge,
        scopesList,
      });
      this.openRedirectUrl(redirect);
    } catch (e) {
      this.error = e.message;
    }
  }

  /**
   * @param {object} params
   * @param {string} params.consentChallenge
   * @return {Promise<{scopesList: string[], skip: boolean}>}
   */
  @Action
  async loadScopes({ consentChallenge }) {
    let scopesList = [];
    let isSkip = false;

    try {
      const {
        requested_scope: requestedScope,
        skip,
        redirect_url: redirectUrl,
      } = await this.accountsStore.getConsentDetails(consentChallenge);

      isSkip = skip;
      scopesList = requestedScope;

      if (isSkip) {
        this.openRedirectUrl(redirectUrl);
      }

      if (!requestedScope) {
        this.error = i18n.t('components.consentProvider.scopesRequired');
      }
    } catch (e) {
      this.error = i18n.t('components.consentProvider.loadScopesError');
    }

    return {
      isSkip,
      scopesList,
    };
  }
}

export default () => createController(ConsentProviderController);
