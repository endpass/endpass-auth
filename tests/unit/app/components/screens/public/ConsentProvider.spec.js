import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import ConsentProvider from '@/components/screens/public/ConsentProvider';
import '@mocks/window';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('ConsentProvider', () => {
  let $router;
  let wrapper;
  let store;
  let storeData;
  let coreModule;
  let accountsModule;

  function createRouter() {
    return {
      history: {
        current: {
          query: {
            consent_challenge: 'foo',
          },
        },
      },
      replace: jest.fn(),
    };
  }

  beforeEach(() => {
    jest.clearAllMocks();

    coreModule = {
      state: {
        loading: false,
      },
    };
    accountsModule = {
      state: {
        isLogin: true,
      },
      actions: {
        getConsentDetails: jest.fn().mockResolvedValue({
          requested_scope: [],
        }),
        grantPermissionsWithOauth: jest.fn(),
      },
    };
    storeData = {
      modules: {
        accounts: accountsModule,
        core: coreModule,
      },
    };
    store = new Vuex.Store(storeData);
    $router = {
      history: {
        current: {
          query: {
            consent_challenge: 'foo',
            scopes: 'foo bar baz',
          },
        },
      },
      replace: jest.fn(),
    };
    wrapper = shallowMount(ConsentProvider, {
      localVue,
      store,
      mocks: {
        $router: createRouter(),
      },
    });
  });

  describe('render', () => {
    it('should correctly render ConsentProvider screen', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should takes query params from current location and assign error if consentChallenge is not in params', async () => {
      $router = createRouter();
      $router.history.current.query = {};

      wrapper = shallowMount(ConsentProvider, {
        localVue,
        store,
        mocks: {
          $router,
        },
      });

      expect(wrapper.vm.error.show).toBe(true);
      expect($router.replace).not.toBeCalled();
    });

    it('should should redirect if consent request provided skip', async () => {
      expect.assertions(1);

      // eslint-disable-next-line
      const redirect_url = 'http://kek.kek';
      accountsModule.actions.getConsentDetails.mockResolvedValue({
        skip: true,
        requested_scope: [],
        redirect_url,
      });
      $router = createRouter();
      wrapper = shallowMount(ConsentProvider, {
        localVue,
        store,
        mocks: {
          $router,
        },
      });
      await wrapper.vm.$nextTick();
      expect(window.location.href).toBe(redirect_url);
    });

    it('should takes query params from current location and makes redirect if consentChallenge is not empty but authorization status is falsy', () => {
      $router = createRouter();
      accountsModule.state.isLogin = false;
      wrapper = shallowMount(ConsentProvider, {
        localVue,
        store,
        mocks: {
          $router,
        },
      });

      expect($router.replace).toBeCalled();
    });

    it('should not do anything on mounting if challengeId is present in query params and user authorized', () => {
      wrapper = shallowMount(ConsentProvider, {
        localVue,
        store,
        mocks: {
          $router,
        },
      });

      expect(wrapper.vm.error).toEqual({
        show: false,
        hint: '',
        description: '',
      });
      expect($router.replace).not.toBeCalled();
    });

    it('should grant permissions on scopes form submit', () => {
      wrapper = shallowMount(ConsentProvider, {
        localVue,
        store,
        mocks: {
          $router,
        },
      });
      wrapper.setData({
        scopesList: ['foo', 'bar', 'baz'],
        isLoading: false,
      });

      wrapper
        .find('scopes-form-stub')
        .vm.$emit('submit', ['foo', 'bar', 'baz']);

      expect(accountsModule.actions.grantPermissionsWithOauth).toBeCalledWith(
        expect.any(Object),
        {
          consentChallenge: 'foo',
          scopesList: ['foo', 'bar', 'baz'],
        },
        undefined,
      );
    });
  });
});
