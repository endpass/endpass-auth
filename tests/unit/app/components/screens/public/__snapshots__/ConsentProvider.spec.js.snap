// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentProvider render should correctly render ConsentProvider screen 1`] = `
<v-frame-stub>
  <scopes-form-stub scopeslist=""></scopes-form-stub>
</v-frame-stub>
`;
